include/oneapi/dpl/pstl/hetero/algorithm_ranges_impl_hetero.h

```diff
@@ -92,20 +92,47 @@ __pattern_walk_n(__hetero_tag<_BackendTag>, _ExecutionPolicy&& __exec, _Function
 }
 
 #if _ONEDPL_CPP20_RANGES_PRESENT
+template <typename _Fun, typename _Proj>
+struct __pattern_for_each_fn
+{
+    _Fun __f;
+    _Proj __proj;
+
+    template <typename _TValue>
+    void
+    operator()(_TValue&& __val) const
+    {
+        std::invoke(__f, std::invoke(__proj, std::forward<_TValue>(__val)));
+    }
+};
+
 //---------------------------------------------------------------------------------------------------------------------
 // pattern_for_each
 //---------------------------------------------------------------------------------------------------------------------
 template <typename _BackendTag, typename _ExecutionPolicy, typename _R, typename _Proj, typename _Fun>
 void
 __pattern_for_each(__hetero_tag<_BackendTag> __tag, _ExecutionPolicy&& __exec, _R&& __r, _Fun __f, _Proj __proj)
 {
-    auto __f_1 =
-        [__f, __proj](auto&& __val) { std::invoke(__f, std::invoke(__proj, std::forward<decltype(__val)>(__val)));};
+    __pattern_for_each_fn<_Fun, _Proj> __f_1{__f, __proj};
 
     oneapi::dpl::__internal::__ranges::__pattern_walk_n(__tag, std::forward<_ExecutionPolicy>(__exec), __f_1,
                                                             oneapi::dpl::__ranges::views::all(std::forward<_R>(__r)));
 }
 
+template <typename _F, typename _Proj>
+struct __pattern_transform_unary_op
```

Comment:
It seems we already have the same (with functionality perspective) type above, on line 96
 -  `struct __pattern_for_each_fn`.

Reply:
Fixed